### PR TITLE
Deploy correct Gateway CRDs before EKS addons

### PR DESCRIPTION
**Pull request by tshaynik**
This PR separates the code deploying Kubernetes Gateway CRDs and the Let's encrypt required by traefik into a separate module of resources to be deployed before eks-addons, and solves a few issues:
- No longer using unnecessary `generate` blocks for these resources
- Explicit dependency of EKS Addons on the resources that must be present beforehand, improving deployment workflow
- Uses `kubectl_path_documents` to deploy multiple K8S resources. Previously, the use of `kubectl_manifest` for the CRDs meant that only the first resource was deployed by terraform (presumably the others were deployed manually)
- Using the correct Gateway CRDS
  - Traefik [requires v1alpha2](https://doc.traefik.io/traefik/providers/kubernetes-gateway/#traefik-kubernetes-with-gateway-api) for Gateway CRDs, not v1beta1, and provides resource definitions [here](https://doc.traefik.io/traefik/providers/kubernetes-gateway/#traefik-kubernetes-with-gateway-api). This was causing deployments to break on the new cluster 

### Summary by CodeRabbit

- New Feature: Added Terraform configurations for deploying Kubernetes resources before EKS Addons. This includes a Let's Encrypt issuer and gateway Custom Resource Definitions (CRDs).
- Refactor: Updated the Terraform source module in `terragrunt.hcl` files and added new dependencies.
- Chore: Specified required versions for Terraform and its providers in `versions.tf`.
- Bug Fix: Enabled the experimental Kubernetes Gateway and updated the hostname annotation under the service section in `eks-addons/terragrunt.hcl`.
- Refactor: Removed generation blocks for "letsencrypt_issuer" and "gateway_crds" from `eks-addons/terragrunt.hcl`.

### CODE REVIEW

File: infra/modules/before_addons/main.tf [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+}
+
+
+resource "kubectl_manifest" "letsencrypt_issuer" {
+  yaml_body = file("letsencrypt_issuer.yaml")
+}
+
+data "kubectl_path_documents" "docs" {
+  pattern = "gateway_crds.yaml"
+}
```
[the remark this answers is not in the document]
Comment (tshaynik) [reply infra/modules/before_addons/main.tf]: The argument must be named pattern. This is fine.

File: infra/modules/before_addons/main.tf [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+resource "kubectl_manifest" "gateway_crds" {
+  for_each  = toset(data.kubectl_path_documents.docs.documents)
+  yaml_body = each.value
+}
```
[the remark this answers is not in the document]
Comment (tshaynik) [reply infra/modules/before_addons/main.tf]: The order of documents does not matter.

File: infra/modules/before_addons/main.tf (hunk 1 above)
[the remark this answers is not in the document]
Comment (tshaynik) [reply infra/modules/before_addons/main.tf]: Replaced with the more appropriate [kubectl_file_documents](https://registry.terraform.io/providers/gavinbunney/kubectl/latest/docs/data-sources/kubectl_file_documents)

File: infra/modules/before_addons/main.tf [hunk 3]
```diff
@@ -0,0 +1,19 @@
+variable "cluster_name" {
+  description = "Name of the Kubernetes cluster"
+  default     = "sample-cluster"
+  type        = string
+}
+
+
+resource "kubectl_manifest" "letsencrypt_issuer" {
+  yaml_body = file("letsencrypt_issuer.yaml")
+}
```
[the remark this answers is not in the document]
Comment (tshaynik) [reply infra/modules/before_addons/main.tf]: added `path.module` for clarity.

**Review by shlevy — APPROVED** [review]

**Pull request merged**